### PR TITLE
[dx11] add graphics pipeline creation & shader module creation

### PR DESCRIPTION
**Pull request by fkaa**
Adds basic pipeline creation and some plumbing for creating/converting from hal to dx11 state. The shader stuff is mostly from the dx12 backend without the root signature stuff, need to investigate how much patching needs to be done on the output (SM 5.0 vs 5.1)

PR checklist:
- [ ] `make` succeeds (on *nix)
- [ ] `make reftests` succeeds
- [ ] tested examples with the following backends:

### CODE REVIEW

File: src/backend/dx11/src/lib.rs [hunk 1]
```diff
@@ -358,6 +360,178 @@ impl Device {
             context
         }
     }
+
+    fn create_rasterizer_state(&self, rasterizer_desc: &pso::Rasterizer) -> Result<ComPtr<d3d11::ID3D11RasterizerState>, pso::CreationError> {
```
Comment (kvark) [inline src/backend/dx11/src/lib.rs]: let's stop extending the device in "lib.rs" and do it in "device.rs" like all other backends do

File: src/backend/dx11/src/conv.rs [hunk 2]
```diff
@@ -102,3 +112,210 @@ pub fn map_rect(rect: &Rect) -> D3D11_RECT {
         bottom: rect.h as _,
     }
 }
+
+pub fn map_topology(primitive: Primitive) -> D3D11_PRIMITIVE_TOPOLOGY {
```
Comment (kvark) [inline src/backend/dx11/src/conv.rs]: another thing to share with dx12? assuming `D3D11_PRIMITIVE_TOPOLOGY == D3D_PRIMITIVE_TOPOLOGY`

File: src/backend/dx11/src/lib.rs (hunk 1 above)
Comment (fkaa) [reply src/backend/dx11/src/lib.rs]: I should probably get rid of the files which are leftover from before too.

Do you mind if I do this on the next PR? I have some commits which are based off of this branch already and can see the merge conflicts happening already in my mind..

File: src/backend/dx11/src/conv.rs (hunk 2 above)
Comment (fkaa) [reply src/backend/dx11/src/conv.rs]: Yup

File: src/backend/dx11/src/lib.rs (hunk 1 above)
Comment (kvark) [reply src/backend/dx11/src/lib.rs]: sure

File: src/backend/dx11/src/lib.rs [hunk 3]
```diff
@@ -358,6 +360,178 @@ impl Device {
...
+        let desc = conv::map_rasterizer_desc(rasterizer_desc);
+
+        let hr = unsafe {
+            self.device.CreateRasterizerState(
+                &desc,
+                &mut rasterizer as *mut *mut _ as *mut *mut _
+            )
+        };
+
+        if !winerror::SUCCEEDED(hr) {
```
Comment (msiglreith) [inline src/backend/dx11/src/lib.rs]: Swap those
```
if SUCCEED { } else { }
```

File: src/backend/dx11/src/lib.rs [hunk 4]
```diff
@@ -358,6 +360,178 @@ impl Device {
...
+        let desc = conv::map_blend_desc(blend_desc);
+
+        let hr = unsafe {
+            self.device.CreateBlendState(
+                &desc,
+                &mut blend as *mut *mut _ as *mut *mut _
+            )
+        };
+
+        if !winerror::SUCCEEDED(hr) {
```
Comment (msiglreith) [inline src/backend/dx11/src/lib.rs]: same here and further down

File: src/backend/dx11/src/lib.rs [hunk 5]
```diff
@@ -415,14 +591,58 @@ impl hal::Device<Backend> for Device {
...
+        let rasterizer_state = self.create_rasterizer_state(&desc.rasterizer)?;
+        let blend_state = self.create_blend_state(&desc.blender)?;
+        let depth_stencil_state = if let Some(desc) = desc.depth_stencil {
+            Some(self.create_depth_stencil_state(&desc)?)
+        } else {
+            None
+        };
+
+        let vs = self.create_vertex_shader(vs)?;
+        let ps = if let Some(blob) = ps {
```
Comment (msiglreith) [inline src/backend/dx11/src/lib.rs]: `ps.map(..)`?

Comment (fkaa) [reply src/backend/dx11/src/lib.rs]: IIRC compiler complained about `?` inside closure or something similar, i'll give it another try

**Comment by fkaa** [comment]
@msiglreith Updated the error handling, but I couldn't get `map()` to work (`.map()` on `Option<T>` has to return `Option<_>`, but the `create_pixel_shader` returns `Result<T, _>` and so on).

I'll revisit this later but I think at least for now it reads more clear than having a series of transform operations on the return value and nested `?` in closures.

**Pull request merged**